File: clients/android/app/src/main/java/network/buildit/modules/fundraising/FundraisingModule.kt
package network.buildit.modules.fundraising

import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AttachMoney
import dagger.Binds
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoSet
import kotlinx.serialization.json.Json
import network.buildit.core.modules.BuildItModule
import network.buildit.core.modules.ModuleRoute
import network.buildit.core.nostr.NostrClient
import network.buildit.core.nostr.NostrEvent
import network.buildit.core.storage.BuildItDatabase
import network.buildit.modules.fundraising.data.FundraisingRepository
import network.buildit.modules.fundraising.data.local.CampaignEntity
import network.buildit.modules.fundraising.data.local.CampaignStatus
import network.buildit.modules.fundraising.data.local.CampaignVisibility
import network.buildit.modules.fundraising.data.local.CampaignsDao
import network.buildit.modules.fundraising.data.local.DonationEntity
import network.buildit.modules.fundraising.data.local.DonationStatus
import network.buildit.modules.fundraising.data.local.DonationsDao
import network.buildit.modules.fundraising.data.local.ExpenseEntity
import network.buildit.modules.fundraising.data.local.ExpensesDao
import network.buildit.modules.fundraising.data.local.PaymentMethod
import network.buildit.modules.fundraising.domain.FundraisingUseCase
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Fundraising module for BuildIt.
 *
 * Provides:
 * - Campaign creation and management
 * - Donation tracking
 * - Expense reporting
 * - Crypto payment integration
 */
class FundraisingModule @Inject constructor(
    private val fundraisingUseCase: FundraisingUseCase,
    private val repository: FundraisingRepository,
    private val nostrClient: NostrClient
) : BuildItModule {
    override val identifier: String = "fundraising"
    override val version: String = "1.0.0"
    override val displayName: String = "Fundraising"
    override val description: String = "Create campaigns and accept donations"

    private var subscriptionId: String? = null

    override suspend fun initialize() {
        // Subscribe to fundraising-related Nostr events
        subscriptionId = nostrClient.subscribe(
            network.buildit.core.nostr.NostrFilter(
                kinds = getHandledEventKinds(),
                since = System.currentTimeMillis() / 1000 - 86400 * 7 // Last 7 days
            )
        )
    }

    override suspend fun shutdown() {
        subscriptionId?.let { nostrClient.unsubscribe(it) }
    }

    override suspend fun handleEvent(event: NostrEvent): Boolean {
        return when (event.kind) {
            FundraisingUseCase.KIND_CAMPAIGN -> {
                handleCampaignEvent(event)
                true
            }
            FundraisingUseCase.KIND_DONATION -> {
                handleDonationEvent(event)
                true
            }
            FundraisingUseCase.KIND_EXPENSE -> {
                handleExpenseEvent(event)
                true
            }
            NostrClient.KIND_DELETE -> {
                handleDeletionEvent(event)
                true
            }
            else -> false
        }
    }

    override fun getNavigationRoutes(): List<ModuleRoute> {
        return listOf(
            ModuleRoute(
                route = "fundraising",
                title = "Fundraising",
                icon = Icons.Default.AttachMoney,
                showInNavigation = true,
                content = { args ->
                    // CampaignsListScreen handled by navigation
                }
            ),
            ModuleRoute(
                route = "fundraising/{campaignId}",
                title = "Campaign Details",
                icon = null,
                showInNavigation = false,
                content = { args ->
                    val campaignId = args["campaignId"] ?: return@ModuleRoute
                    // CampaignDetailScreen handled by navigation
                }
            ),
            ModuleRoute(
                route = "fundraising/create",
                title = "Create Campaign",
                icon = null,
                showInNavigation = false,
                content = { args ->
                    // CreateCampaignScreen handled by navigation
                }
            ),
            ModuleRoute(
                route = "fundraising/{campaignId}/donate",
                title = "Donate",
                icon = null,
                showInNavigation = false,
                content = { args ->
                    val campaignId = args["campaignId"] ?: return@ModuleRoute
                    // DonateScreen handled by navigation
                }
            )
        )
    }

    override fun getHandledEventKinds(): List<Int> {
        return listOf(
            FundraisingUseCase.KIND_CAMPAIGN,
            FundraisingUseCase.KIND_DONATION,
            FundraisingUseCase.KIND_EXPENSE,
            NostrClient.KIND_DELETE
        )
    }

    /**
     * Handles incoming campaign events from Nostr.
     */
    private suspend fun handleCampaignEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val campaign = CampaignEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                title = content["title"]?.toString()?.trim('"') ?: return,
                description = content["description"]?.toString()?.trim('"'),
                goal = content["goal"]?.toString()?.toDoubleOrNull() ?: return,
                currency = content["currency"]?.toString()?.trim('"') ?: "USD",
                raised = content["raised"]?.toString()?.toDoubleOrNull() ?: 0.0,
                donorCount = content["donorCount"]?.toString()?.toIntOrNull() ?: 0,
                startsAt = content["startsAt"]?.toString()?.toLongOrNull(),
                endsAt = content["endsAt"]?.toString()?.toLongOrNull(),
                status = CampaignStatus.fromValue(
                    content["status"]?.toString()?.trim('"') ?: "draft"
                ),
                visibility = CampaignVisibility.fromValue(
                    content["visibility"]?.toString()?.trim('"') ?: "group"
                ),
                groupId = nostrEvent.tags.find { it.firstOrNull() == "g" }?.getOrNull(1),
                image = content["image"]?.toString()?.trim('"'),
                tiersJson = content["tiers"]?.toString(),
                updatesJson = content["updates"]?.toString(),
                createdBy = content["createdBy"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                createdAt = content["createdAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt,
                updatedAt = content["updatedAt"]?.toString()?.toLongOrNull()
            )

            repository.saveCampaign(campaign)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle campaign event", e)
        }
    }

    /**
     * Handles incoming donation events from Nostr.
     */
    private suspend fun handleDonationEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val donation = DonationEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                campaignId = content["campaignId"]?.toString()?.trim('"') ?: return,
                amount = content["amount"]?.toString()?.toDoubleOrNull() ?: return,
                currency = content["currency"]?.toString()?.trim('"') ?: "USD",
                donorPubkey = content["donorPubkey"]?.toString()?.trim('"'),
                donorName = content["donorName"]?.toString()?.trim('"'),
                anonymous = content["anonymous"]?.toString() == "true",
                message = content["message"]?.toString()?.trim('"'),
                tierId = content["tierId"]?.toString()?.trim('"'),
                paymentMethod = PaymentMethod.fromValue(
                    content["paymentMethod"]?.toString()?.trim('"') ?: "other"
                ),
                status = DonationStatus.fromValue(
                    content["status"]?.toString()?.trim('"') ?: "completed"
                ),
                donatedAt = content["donatedAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt
            )

            repository.saveDonation(donation)

            // Update campaign totals if donation is completed
            if (donation.status == DonationStatus.COMPLETED) {
                repository.addDonationToCampaign(donation.campaignId, donation.amount)
            }
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle donation event", e)
        }
    }

    /**
     * Handles incoming expense events from Nostr.
     */
    private suspend fun handleExpenseEvent(nostrEvent: NostrEvent) {
        try {
            val json = Json { ignoreUnknownKeys = true }
            val content = json.decodeFromString<Map<String, kotlinx.serialization.json.JsonElement>>(nostrEvent.content)

            val expense = ExpenseEntity(
                id = content["id"]?.toString()?.trim('"') ?: return,
                schemaVersion = content["_v"]?.toString()?.trim('"') ?: "1.0.0",
                campaignId = content["campaignId"]?.toString()?.trim('"') ?: return,
                amount = content["amount"]?.toString()?.toDoubleOrNull() ?: return,
                currency = content["currency"]?.toString()?.trim('"') ?: "USD",
                description = content["description"]?.toString()?.trim('"') ?: return,
                category = content["category"]?.toString()?.trim('"'),
                receipt = content["receipt"]?.toString()?.trim('"'),
                vendor = content["vendor"]?.toString()?.trim('"'),
                date = content["date"]?.toString()?.toLongOrNull(),
                recordedBy = content["recordedBy"]?.toString()?.trim('"') ?: nostrEvent.pubkey,
                recordedAt = content["recordedAt"]?.toString()?.toLongOrNull() ?: nostrEvent.createdAt
            )

            repository.saveExpense(expense)
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle expense event", e)
        }
    }

    /**
     * Handles deletion events from Nostr.
     */
    private suspend fun handleDeletionEvent(nostrEvent: NostrEvent) {
        try {
            val campaignIds = nostrEvent.tags.filter { it.firstOrNull() == "e" }
                .mapNotNull { it.getOrNull(1) }

            campaignIds.forEach { campaignId ->
                repository.deleteCampaign(campaignId)
            }
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to handle deletion event", e)
        }
    }

    companion object {
        private const val TAG = "FundraisingModule"
    }
}

/**
 * Hilt module for Fundraising dependencies.
 */
@Module
@InstallIn(SingletonComponent::class)
abstract class FundraisingHiltModule {
    @Binds
    @IntoSet
    abstract fun bindFundraisingModule(impl: FundraisingModule): BuildItModule
}

/**
 * Provides DAOs for Fundraising module.
 */
@Module
@InstallIn(SingletonComponent::class)
object FundraisingDaoModule {
    @Provides
    @Singleton
    fun provideCampaignsDao(database: BuildItDatabase): CampaignsDao = database.campaignsDao()

    @Provides
    @Singleton
    fun provideDonationsDao(database: BuildItDatabase): DonationsDao = database.donationsDao()

    @Provides
    @Singleton
    fun provideExpensesDao(database: BuildItDatabase): ExpensesDao = database.expensesDao()
}

File: clients/ios/BuildIt/Modules/Fundraising/FundraisingModule.swift
// FundraisingModule.swift
// BuildIt - Decentralized Mesh Communication
//
// Fundraising module for campaigns, donations, and financial tracking.

import Foundation
import SwiftUI
import os.log

/// Fundraising module implementation
@MainActor
public final class FundraisingModule: BuildItModule {
    // MARK: - Module Metadata

    public static let identifier = "fundraising"
    public static let version = "1.0.0"
    public static let dependencies: [String] = []

    // MARK: - Properties

    private let store: FundraisingStore
    private let service: FundraisingService
    private let configManager = ModuleConfigurationManager.shared
    private let logger = Logger(subsystem: "com.buildit", category: "FundraisingModule")

    // MARK: - Initialization

    public init() throws {
        self.store = try FundraisingStore()
        self.service = FundraisingService(store: store)
        logger.info("Fundraising module created")
    }

    // MARK: - BuildItModule Implementation

    public func initialize() async throws {
        logger.info("Initializing Fundraising module")

        // Enable by default for global scope
        try await enable(for: nil)

        logger.info("Fundraising module initialized")
    }

    public func handleEvent(_ event: NostrEvent) async {
        // Route fundraising-related Nostr events to service
        await service.processNostrEvent(event)
    }

    public func getViews() -> [ModuleView] {
        [
            ModuleView(
                id: "fundraising",
                title: "Fundraising",
                icon: "dollarsign.circle",
                order: 30
            ) {
                CampaignsListView(service: service)
            }
        ]
    }

    public func cleanup() async {
        logger.info("Cleaning up Fundraising module")
    }

    public func isEnabled(for groupId: String?) -> Bool {
        configManager.isModuleEnabled(Self.identifier, for: groupId)
    }

    public func enable(for groupId: String?) async throws {
        configManager.enableModule(Self.identifier, for: groupId)
        logger.info("Enabled Fundraising module for group: \(groupId ?? "global")")
    }

    public func disable(for groupId: String?) async {
        configManager.disableModule(Self.identifier, for: groupId)
        logger.info("Disabled Fundraising module for group: \(groupId ?? "global")")
    }

    // MARK: - Public API

    /// Create a new campaign
    public func createCampaign(
        title: String,
        description: String?,
        goal: Double,
        currency: String = "USD",
        startsAt: Date? = nil,
        endsAt: Date? = nil,
        visibility: CampaignVisibility = .group,
        groupId: String? = nil,
        image: String? = nil,
        tiers: [DonationTier] = [],
        cryptoPayment: CryptoPaymentInfo? = nil
    ) async throws -> Campaign {
        try await service.createCampaign(
            title: title,
            description: description,
            goal: goal,
            currency: currency,
            startsAt: startsAt,
            endsAt: endsAt,
            visibility: visibility,
            groupId: groupId,
            image: image,
            tiers: tiers,
            cryptoPayment: cryptoPayment
        )
    }

    /// Record a donation
    public func recordDonation(
        campaignId: String,
        amount: Double,
        currency: String = "USD",
        donorName: String? = nil,
        anonymous: Bool = false,
        message: String? = nil,
        paymentMethod: PaymentMethod = .crypto,
        cryptoType: CryptoType? = nil
    ) async throws -> Donation {
        try await service.recordDonation(
            campaignId: campaignId,
            amount: amount,
            currency: currency,
            donorName: donorName,
            anonymous: anonymous,
            message: message,
            paymentMethod: paymentMethod,
            cryptoType: cryptoType
        )
    }

    /// Get all campaigns
    public func getCampaigns(
        groupId: String? = nil,
        activeOnly: Bool = false
    ) async throws -> [Campaign] {
        try await service.getCampaigns(groupId: groupId, activeOnly: activeOnly)
    }

    /// Get donations for a campaign
    public func getDonations(campaignId: String) async throws -> [Donation] {
        try await service.getDonations(campaignId: campaignId)
    }

    /// Get campaign analytics
    public func getAnalytics(for campaignId: String) async throws -> CampaignAnalytics {
        try await service.calculateAnalytics(for: campaignId)
    }
}
